spec/java/util/stream/Stream.main.lsl

```diff
@@ -0,0 +1,1049 @@
+                storage = this.storage,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        else
+        {
+            // plain bubble sorting algorithm
+            val outerLimit: int = this.length - 1;
+            var innerLimit: int = 0;
+            var i: int = 0;
+            var j: int = 0;
+            action LOOP_FOR(
+                i, 0, outerLimit, +1,
+                sort_loop_outer_with_comparator(i, j, innerLimit, comparator)
+            );
+
+            result = new StreamAutomaton(state = Initialized,
+                storage = this.storage,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc sort_loop_outer_with_comparator (i: int, j: int, innerLimit: int, comparator: Comparator): void
+    {
+        innerLimit = this.length - i - 1;
+        action LOOP_FOR(
+            j, 0, innerLimit, +1,
+            sort_loop_inner_with_comparator(j, comparator)
+        );
+    }
+
+
+    @Phantom proc sort_loop_inner_with_comparator (j: int, comparator: Comparator): void
+    {
+        val idxA: int = j;
+        val idxB: int = j + 1;
+        val a: Object = this.storage[idxA];
+        val b: Object = this.storage[idxB];
+
+        // "compareTo" - that's right ??
+        if (action CALL(comparator, [a, b]) > 0)
+        {
+            this.storage[idxA] = b;
+            this.storage[idxB] = a;
+        }
+    }
+
+
+    fun *.peek (@target self: Stream, _action: Consumer): Stream
+    {
+        result = _actionApply(_action);
+    }
+
+
+    // maxSize: long - what to do with this ? Array
+    fun *.limit (@target self: Stream, maxSize: long): Stream
+    {
+        if (maxSize < 0)
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+
+        if (maxSize == 0)
+        {
+            result = new StreamAutomaton(state = Initialized,
+                storage = this.storage,
+                length = 0,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        // Maybe only change length field ? And don't change storage ?
+        else
+        {
+            // what will be if will be overflow ?
+            val maxSizeInt: int = maxSize as int;
+            val limitStorage: array<Object> = action ARRAY_NEW("java.lang.Object", maxSizeInt);
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, maxSizeInt, +1,
+                _limit_loop(i, limitStorage)
+            );
+
+            result = new StreamAutomaton(state = Initialized,
+                storage = limitStorage,
+                length = maxSizeInt,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc _limit_loop (i: int, limitStorage: array<Object>): void
+    {
+        limitStorage[i] = this.storage[i];
```

Comment:
@dog-m In original method willn't be any exception; It will return all elements;

```
        Stream<Integer> integerStream1 = Stream.of(5, 3, 6);

        integerStream1.limit(7).forEach(System.out::println);
```

Reply:
Fixed